crates/oxc_ast/src/syntax_directed_operations/gather_node_parts.rs

```diff
@@ -41,8 +41,10 @@ impl<'a> GatherNodeParts<'a> for MemberExpression<'a> {
 impl<'a> GatherNodeParts<'a> for AssignmentTarget<'a> {
     fn gather<F: FnMut(Atom<'a>)>(&self, f: &mut F) {
         match self {
-            AssignmentTarget::SimpleAssignmentTarget(t) => t.gather(f),
-            AssignmentTarget::AssignmentTargetPattern(_) => {}
+            match_simple_assignment_target!(Self) => {
```

Comment:
The only part I don't like about this is these match patterns, using this much macro gives me C PTSD; it is a bit less hygienic than what I would prefer. Otherwise, it looks good to me.

Reply:
@overlookmotel Is there any specific case that these match macros solve? I'm having a hard time figuring out what we are trying to achieve with this syntax sugar. I think `x.is_simple_assignment_target` is a much better interface to work with Since we won't have to know an obscure macro. In some cases we might want to write the variants manually instead of using `is_*` methods but the user should decide and I believe if they understand the `@inherit` then they would have no issues using it.

On another note, Please consider using `codegen` not as a tool that makes writing code simpler but as a tool for documentation since I most of the time find myself reading the source code instead of documentation and it is my biggest issue with complex macros, It makes following the code a bit obfuscated.

Reply:
Having the definition at my fingertips is a superpower even AI tools wouldn't match. Comments get outdated but the code is always valid. Seeing a macro definition when I'm looking for the code is almost always a turn-down to me since I know it wouldn't be a simple code and I have to spend a few minutes just figuring out the macro and creating a mental model of what it looks like at the runtime.

This isn't really true for small macros, But as the macro complexity expands so does the time a new developer needs to have a good mental model when using it.

Reply:
Thanks for reviewing.

Yes, it's not ideal. There is no syntax for matching against multiple variants without either a macro, or writing out all the variants by hand. In the case of `Expression`, obviously the latter is impractical.

The alternative which does work is:

```rs
match self {
    _ if self.is_simple_assignment_target() => { }
    /* ... */
}
```

But problem with that is it doesn't count as an exhaustive match.

I also suspect the macro (which expands to all the variants) may help the compiler to optimize the match, as that syntax allows the match arms to be re-ordered, so I used it everywhere, even where exhaustive matching isn't required and `_ if self.is_simple_assignment_target()` would work.

I tried to find a better way, but only thing I could think of was `match!` macro, but that seemed too much.

Swings and roundabouts - in other places the syntax is nicer/shorter, but in return the `match` syntax is not great.

Reply:
Ah, I see, Now I get your reasoning behind this.

I would love to see the performance difference between this macro and having the `is_*` methods `#[inline(always]`, I'm hoping that the compiler treats them the same after inlining the function call.

By the way, Great work! There is a lot of thought behind this.

Reply:
Above `match!` example would expand to:

```rs
match self {
    _temp @ match_expression!(Self) => {
        let expr = _temp.to_expression();
        expr.do_stuff()
    }
}
```

Reply:
> Above `match!` example would expand to:
> 
> ```rust
> match self {
>     _temp @ match_expression!(Self) => {
>         let expr = _temp.to_expression();
>         expr.do_stuff()
>     }
> }
> ```

FYI, I have no problem with the syntax of the first example(without `match!`), What I'm trying to say is that if the performance penalty of using methods over macros is small we might as well do it that way to avoid having these highly specialized match macros all over the place.

Reply:
The `match!` macro idea is growing on me. Perhaps if the naming was more explicit, it would be less magical?

```rs
match_with_subtypes!(prop_key, {
    PropertyKey::StaticIdentifier(id) => id.do_stuff(),
    PropertyKey::PrivateIdentifier(id) = id.do_stuff(),
    PropertyKey::@subtype::Expression(expr) => expr.do_stuff(),
})
```

I know this goes in the opposite direction of your original comment - that you want less macros not more! I'm just thinking of being able to write tight expressive code without boilerplate.

> I would love to see the performance difference between this macro and having the is_* methods #[inline(always], I'm hoping that the compiler treats them the same after inlining the function call.

I haven't tried it. Feel free to measure if you like. `_ if self.is_expression()` might be just as fast (even without `#[inline_always]` - the compiler is very aggressive in inlining small functions anyway). To find out would be mostly just be find-and-replace.

However, `_ if self.is_expression()` doesn't work in an exhaustive match, which personally I think is a major downside. When your `match` has no catch-all `_ => {}` arm, it guarantees you've handled every case. If another variant is added to the enum later, then when using exhaustive matching, the compiler will flag everywhere where you need to handle this new case. Not so if using `_ if self.is_expression()`.

Reply:
I get where you are coming from, And I totally understand your concerns mentioned above.

I just think there are not that many cases that need to match all of the expression(or statement) variants + some other stuff. It usually boils down to a few items, And where we need to actually match the whole thing we can always fall back on the `is_expression` and `as_expression` methods. I've checked the changes in this PR and couldn't really find a use case with big enum types, Could you please point me toward one of the necessary examples that requires a bunch of boilerplate for the match?

Reply:
There are 2 parts to these macros, One side is the generation which is done with macros like `inherit_variants` I strongly think these should be done with a codegen or at least a `proc-macro` - yes although it is slower to build it is more friendly with tools - As it is right now it totally messes up the LSP and `rust-analyzer` and things like jump to definition, documentation and completion-suggestions are broken in a bunch of scenarios.

As for the match macros, I think we can live with or without it. Most cases can be written manually (except the `Statement` and `Expression` enums which are huge) so maybe we can limit the scope of these to bigger types?
To be honest I think the match macros can be used without any issues(other than the learning curve) as long as we are generating the AST so we can have the code and tooling around it.

If we generate everything, Then we only have match macros to worry about and they don't do `@inherit`/`@subtype` magic, They are just copying the boilerplate for you, Which is a totally valid use-case in my opinion; If it's not right for this then what is for after all?

Reply:
I've marked this as unresolved just so can see and don't forget these considerations.